features/users/controller.js

```diff
@@ -0,0 +1,152 @@
+var Chance = new require('chance')();
+
+var User = require('./model');
+var Friendship = require('./friendship.service');
+var BaseApi = require("../../api");
+
+var mocks = require('../../test-helpers/_mocks');
+
+class UsersController extends BaseApi {
+
+  async getUserById(id) {
+    try {
+      let user = await User.findById(id).select(['-friends']);
+      let friends = await Friendship.getFriends(id);
+      return {
+        user: user,
+        friends: friends
+      };
+    } catch(ex) {
+      throw ex;
+    }
+  }
+
+  getUser(req, res) {
+    let id = req.params.id;
+    log.info("looking for user id " + id);
+    this.getUserById(id)
+      .then(model => this.success(res, model))
+      .catch(err => this.exception(res, err));
+  }
+
+  async getUsers(req, res) {
+    await User.find()
+      .select(['-friends'])
+      .sort({ "name": 1 })
+      .exec( (err, models) => {
+        if (err) {
+          log.error(err);
+          return this.exception(res, err);
+        }
+        return this.success(res, models);
+      });
+  };
+
+  async getMe(req, res) {
+    let data = this.getToken(req);
+    log.info(data);
+
+    try {
+      let user = await this.getUserById(data.user_id);
+      if (!user) {
+        return this.fail(res, 401, "User not found!");
+      }
+
+      return this.success(res, user);
+    } catch(ex) {
+      return this.exception(res, ex);
+    }
+  }
+
+  createUser(req, res) {
+    var model = new User(req.body);
+    return new Promise((resolve, reject) => {
+      model.save((err, model) => {
+        if (err) {
+          log.error("save user error", err);
+          this.exception(res, err);
+          return reject(err);
+        }
+        this.success(res, model);
+        return resolve(model);
+      });
+    })
+  }
+
+  async meet(req, res) {
+    let user = req.params.id;
+    let friend = req.params.friend;
+    try {
+      await Friendship.makeFriend(user, friend);
+      let response = await this.getUserById(user);
+      return this.success(res, response);
+    } catch(ex) {
+      return this.exception(res, ex);
+    }
+  }
+  async breakUp(req, res) {
+    let user = req.params.id;
+    let friend = req.params.friend;
+    log.info(user + " breaking up with "+ friend);
+    try {
+      await Friendship.breakUp(user, friend);
+      let response = await this.getUserById(user);
+      return this.success(res, response);
+    } catch(ex) {
+      return this.exception(res, ex);
+    }
+  }
+
+  updateUser(req, res) {
+    var id = req.params.id;
+    let data = this.getToken(req);
+    if(!data || data.user_id != id) return this.error(res, 401, "Denied");
+
+    User.update({ _id: id }, req.body)
+      .then((model) => {
+        this.getUserById(id)
+          .then(model => this.success(res, model))
+          .catch(err => this.exception(res, err));
+      })
+      .catch(err => this.exception(res, err));
+  }
+
+  async deleteUser(req, res) {
```

Comment:
Also I noticed if I deleted an user, then get the user by Id, in stead of return 404, it return 500 with empty data:
```
curl -XDELETE 'http://localhost:3003/user/5d67841dcb1ac123d747ddbd'   
{"success":true,"data":{"friends":[],"_id":"5d67841dcb1ac123d747ddbd","name":"banddd","email":"d@t.com","createdAt":"2019-08-29T07:51:57.867Z","updatedAt":"2019-08-29T07:51:57.867Z","__v":0}}%    

curl -v -XGET 'http://localhost:3003/user/5d67841dcb1ac123d747ddbd'                                                                                  
*   Trying ::1...
* Connected to localhost (::1) port 3003 (#0)
> GET /user/5d67841dcb1ac123d747ddbd HTTP/1.1
> Host: localhost:3003
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< X-Powered-By: Express
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: PUT, GET, POST, DELETE, OPTIONS
< Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Client, Access-Control-Allow-Credentials
< Access-Control-Allow-Credentials: true
< Content-Type: application/json; charset=utf-8
< Content-Length: 27
< ETag: W/"1b-0kBWvE2JG0rfOkAnDSV0ZoE7X28"
< Date: Thu, 29 Aug 2019 09:06:09 GMT
< Connection: keep-alive
< 
* Connection #0 to host localhost left intact
{"success":false,"data":{}}%     
```
Any reason you design it this way?

Reply:
this is an error from my part. I don't have a function in `api.js` for _NOT FOUND_ errors.

the ideal solution would be to create this function and, everywhere I could get data from database, I check the exception to see if we're dealing with an actual error or a 404.